File: example/InstagramClient/InstagramClient/IGViewController.m
//
//  IGViewController.m
//  InstagramClient
//
//  Created by Cristiano Severini on 12/07/12.
//  Copyright (c) 2012 Crino. All rights reserved.
//

#import "IGViewController.h"

#import "IGAppDelegate.h"
#import "IGListViewController.h"

@interface IGViewController ()

@end

@implementation IGViewController

- (void)viewDidLoad {
    [super viewDidLoad];
	// Do any additional setup after loading the view, typically from a nib.
    self.view.backgroundColor = [UIColor whiteColor];
    
    UIButton* loginButton = [UIButton buttonWithType:UIButtonTypeRoundedRect];
    [loginButton setTitle:@"Login" forState:UIControlStateNormal];
    [loginButton sizeToFit];
    loginButton.center = CGPointMake(160, 200);
    [loginButton addTarget:self 
                    action:@selector(login) 
          forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:loginButton];
    
    
    IGAppDelegate* appDelegate = (IGAppDelegate*)[UIApplication sharedApplication].delegate;
    
    // here i can set accessToken received on previous login 
    appDelegate.instagram.accessToken = [[NSUserDefaults standardUserDefaults] objectForKey:@"accessToken"];
    appDelegate.instagram.sessionDelegate = self;
    if ([appDelegate.instagram isSessionValid]) {
        IGListViewController* viewController = [[IGListViewController alloc] init];
        [self.navigationController pushViewController:viewController animated:YES];
    } else {
        [appDelegate.instagram authorize:[NSArray arrayWithObjects:@"comments", @"likes", nil]];
    }
}

- (void)viewDidUnload {
    [super viewDidUnload];
    // Release any retained subviews of the main view.
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation {
    return (interfaceOrientation != UIInterfaceOrientationPortraitUpsideDown);
}

-(void)login {
    IGAppDelegate* appDelegate = (IGAppDelegate*)[UIApplication sharedApplication].delegate;
    [appDelegate.instagram authorize:[NSArray arrayWithObjects:@"comments", @"likes", nil]];
}

#pragma - IGSessionDelegate

-(void)igDidLogin {
    NSLog(@"Instagram did login");
    // here i can store accessToken
    IGAppDelegate* appDelegate = (IGAppDelegate*)[UIApplication sharedApplication].delegate;
    [[NSUserDefaults standardUserDefaults] setObject:appDelegate.instagram.accessToken forKey:@"accessToken"];
	[[NSUserDefaults standardUserDefaults] synchronize];
    
    IGListViewController* viewController = [[IGListViewController alloc] init];
    [self.navigationController pushViewController:viewController animated:YES];
}

-(void)igDidNotLogin:(BOOL)cancelled {
    NSLog(@"Instagram did not login");
    NSString* message = nil;
    if (cancelled) {
        message = @"Access cancelled!";
    } else {
        message = @"Access denied!";
    }
    UIAlertView* alertView = [[UIAlertView alloc] initWithTitle:@"Error"
                                                        message:message
                                                       delegate:nil
                                              cancelButtonTitle:@"Ok"
                                              otherButtonTitles:nil];
    [alertView show];
}

-(void)igDidLogout {
    NSLog(@"Instagram did logout");
    // remove the accessToken
    [[NSUserDefaults standardUserDefaults] setObject:nil forKey:@"accessToken"];
	[[NSUserDefaults standardUserDefaults] synchronize];
}

-(void)igSessionInvalidated {
    NSLog(@"Instagram session was invalidated");
}

@end

File: example/InstagramRubyMotionClient/app/IGViewController.rb
#
# IGViewController.rb
# InstagramClient
#
# Created by Devon Blandin on 03/25/13
# http://hello.devonblandin.com
#

class IGViewController < UIViewController

  def viewDidLoad
    super

    self.view.backgroundColor = UIColor.whiteColor

    login_button = UIButton.buttonWithType(UIButtonTypeRoundedRect)
    login_button.setTitle('Login', forState: UIControlStateNormal)
    login_button.sizeToFit
    login_button.center = CGPointMake(160, 200)
    login_button.addTarget(self, action: 'login', forControlEvents: UIControlEventTouchUpInside)

    self.view.addSubview(login_button)

    App.delegate.instagram.accessToken = NSUserDefaults.standardUserDefaults['accessToken']
    App.delegate.instagram.sessionDelegate = self

    if App.delegate.instagram.isSessionValid
      view_controller = IGListViewController.alloc.init
      self.navigationController.pushViewController(view_controller, animated: true)
    else
      App.delegate.instagram.authorize(['comments', 'likes'])
    end
  end

  def viewDidUnload
    super
  end

  def shouldAutorotateToInterfaceOrientation(interfaceOrientation)
    interfaceOrientation != UIInterfaceOrientationPortraitUpsideDown
  end

  def login
    App.delegate.instagram.authorize(['comments', 'likes'])
  end

  def igDidLogin
    NSUserDefaults.standardUserDefaults['accessToken'] = App.delegate.instagram.accessToken
    NSUserDefaults.standardUserDefaults.synchronize

    view_controller = IGListViewController.alloc.init
    self.navigationController.pushViewController(view_controller, animated: true)
  end

  def igDidNotLogin(cancelled)
    NSLog('Instagram did not login')

    message = cancelled ? 'Access cancelled!' : 'Access denied!'
    alert_view = UIAlertView.alloc.initWithTitle('Error',
                                                 message: message,
                                                delegate: nil,
                                       cancelButtonTitle: 'Ok',
                                       otherButtonTitles: nil)
    alert_view.show
  end

  def igDidLogout
    NSLog('Instagram did logout')

    NSUserDefaults.standardUserDefaults['accessToken'] = nil
    NSUserDefaults.standardUserDefaults.synchronize
  end

  def igSessionInvalidated
    NSLog('Instagram session was invalidated')
  end
end
